products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -25,96 +25,127 @@ using namespace llbuild::buildsystem;
+class CAPIBuildDB: public BuildDBDelegate {
+  /// The key table, for memory caching of key to key id mapping.
+  llvm::StringMap<KeyID> keyTable;
+  
+  /// The mutex that protects the key table.
+  std::mutex keyTableMutex;
+  
+  std::unique_ptr<BuildDB> _db;
+  
+  /// The mutex that protects the key cache.
+  std::mutex keyCacheMutex;
+  
+  std::unordered_map<KeyID, CAPIBuildKey *> keyCache;
+  
+  CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
+    _db = createSQLiteBuildDB(path, clientSchemaVersion, /* recreateUnmatchedVersion = */ false, error_out);
+  }
+  
+  bool fetchKeysIfNecessary(std::string *error) {
+    if (keyCache.size() > 0) {
+      return true;
     }
+    std::vector<KeyType> keys;
     
-    llb_build_key_t *keyAtIndex(int32_t index) {
-      auto buildKey = BuildKey::fromData(keys[index]);
-      return (llb_build_key_t *)new CAPIBuildKey(buildKey);
-    }
-  };
-
-  class CAPIBuildDB: public BuildDBDelegate {
-    /// The key table, for memory caching of key to key id mapping.
-    llvm::StringMap<KeyID> keyTable;
-    
-    /// The mutex that protects the key table.
-    std::mutex keyTableMutex;
-    
-    std::unique_ptr<BuildDB> _db;
+    auto success = _db->getKeys(keys, error);
     
-    CAPIBuildDB(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
-      _db = createSQLiteBuildDB(path, clientSchemaVersion, /* recreateUnmatchedVersion = */ false, error_out);
+    if (!error->empty()) {
+      return success;
     }
     
-  public:
-    static CAPIBuildDB *create(StringRef path, uint32_t clientSchemaVersion, std::string *error_out) {
-      auto databaseObject = new CAPIBuildDB(path, clientSchemaVersion, error_out);
-      if (databaseObject->_db == nullptr || !error_out->empty() || !databaseObject->buildStarted(error_out)) {
-        delete databaseObject;
-        return nullptr;
-      }
-      
-      // The delegate  caches keys for key ids
-      databaseObject->_db.get()->attachDelegate(databaseObject);
-      
-      return databaseObject;
-    }
+    buildUpKeyCache(keys);
     
-    bool isValid(std::string *error_out) {
-      if (_db == nullptr) {
-        return false;
-      }
-      
-      return buildStarted(error_out);
-    }
+    return success;
+  }
+  
+  void buildUpKeyCache(std::vector<KeyType> &allKeys) {
+    std::lock_guard<std::mutex> guard(keyCacheMutex);
     
-    virtual const KeyID getKeyID(const KeyType& key) override {
-      std::lock_guard<std::mutex> guard(keyTableMutex);
-      
-      // The RHS of the mapping is actually ignored, we use the StringMap's ptr
-      // identity because it allows us to efficiently map back to the key string
-      // in `getRuleInfoForKey`.
-      auto it = keyTable.insert(std::make_pair(key, 0)).first;
-      return (KeyID)(uintptr_t)it->getKey().data();
+    keyCache = std::unordered_map<KeyID, CAPIBuildKey *>();
+    keyCache.reserve(allKeys.size());
+    int index = 0;
```

Comment:
Was left over from a previous iteration where I used a vector instead of an unordered_map as the cache. That's not possible though to support gaps in the identifiers. I removed the variable.